core/templates/components/common-layout-directives/navigation-bars/top-navigation-bar.directive.html

```diff
@@ -49,7 +49,7 @@
           <li class="nav-item"> <a ng-keydown="$ctrl.onMenuKeypress($event, 'aboutMenu', {tab: $ctrl.ACTION_CLOSE})" class="dropdown-item nav-link oppia-navbar-tab-content" href="https://oppiafoundation.org/volunteer" target="_blank" rel="noopener" translate="I18N_TOPNAV_GET_INVOLVED"></a></li>
         </ul>
       </li>
-      <li ng-if="SHOW_CLASSROOM_CALLOUT" ng-class="{'open' : $ctrl.activeMenuName === 'classroomMenu'}" uib-dropdown ng-show="$ctrl.navElementsVisibilityStatus.I18N_TOPNAV_CLASSROOM" aria-hidden="false" class="nav-item  oppia-navbar-clickable-dropdown protractor-test-classroom-oppia-list-item">
+      <li ng-class="{'open' : $ctrl.activeMenuName === 'classroomMenu'}" uib-dropdown ng-show="$ctrl.navElementsVisibilityStatus.I18N_TOPNAV_CLASSROOM && ($ctrl.CLASSROOM_PAGE_IS_SHOWN || $ctrl.updateShowClassroom())" aria-hidden="false" class="nav-item  oppia-navbar-clickable-dropdown protractor-test-classroom-oppia-list-item">
```

Comment:
The reason for calling the update function here is that calling it in the `onInit` method leads to some errors. While registering, the session would time out and this behavior was deterministic.

Reply:
You can't do it here though, it doesn't make sense semantically. How about using ng-if instead of ng-show?

Reply:
Hmm, would it be okay if I did the same thing in the ng-if? Also, I tried using ng-init but the same issue recured.

Reply:
I won't suggest not/never adding a method in a digest cycle which makes API call. 

> method leads to some errors.

What error do you exactly see? try not to ignore things which is happening unexpected, it can lead to a bigger problem later.

> While registering, the session would time out and this behavior was deterministic.

Have you tried disabling `REDIRECT_UNFINISHED_SIGNUPS` in the handler?